### DISCUSSION TITLE
select! pattern matching with exclusive references in async expr

### DISCUSSION THREAD

**Discussion by coljnr9** [tokio-rs/tokio#6633] [Q&A] [answered]
I find that I frequently want to `select!` over expressions that 
- Return pattern-matchable values and
- Require exclusive references to _something_

When receiving from a `UdpSocket`, I'd love to do something like

```rust
let mut buf = [0; 64];
select! {
    Ok(n) = socket.recv(&mut buf) => { /* Handle successful rx */ },
    Err(e) = socket.recv(&mut buf) => { /* Handle rx error */ },
     _ = another_future => { /* Handle that */ }
}

```
but I obviously get an error about the multiple borrow of `buf`

similarly, when receiving from an `mpsc::Receiver`
```rust
select! {
    Some(msg) = receiver.recv() => { /* Handle message */ },
    None = receiver.recv() =>  { /* Handle channel closed */ },
    _ =>  another_future => { /* Handle it */ }
}
```
And here I get the (expected) error that `receiver` is mutably borrowed twice.

Right now I handle this by doing the matching outside of the `select!`, but I'm curious if there's a trick to use something like the above syntax.

Thanks!

**Comment by Darksonn** [comment, marked as answer]
Select does not support that, you'll have to match separately from the select.
